feat(forms): add addValidators and hasValidators methods to AbstractControl

dylhunn opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #13461

It is currently not possible to modify the validator set on a FormControl, or check for a particular validator's presence.


## What is the new behavior?

Add getValidators and hasValidators methods (for both sync and async) to the FormControl API, enabling vaioud use cases in a longstanding issue.

Several new functionalities are possible with this change: the most requested is that callers can now check whether a control has a required validator. Other uses include incrementally changing the validators set without doing an expensive operation to reset all validators.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Closes #13461.

dylhunn committed:
feat(forms): add hasValidators, addValidators, and removeValidators methods (for both sync and async)

Several new functionalities are possible with this change: the most requested is that callers can now check whether a control has a required validator. Other uses include incrementally changing the validators set without doing an expensive operation to reset all validators.

Closes #13461.

splincode commented on packages/forms/src/model.ts:
```diff
@@ -444,9 +451,134 @@ export abstract class AbstractControl {
+   *
+   * @param validators The new validator function or functions to add to this control.
+   */
+  addValidators(validators: ValidatorFn|ValidatorFn[]): void {
```
`setValidators` and `addValidators` will confuse the developer, because the meaning is almost the same

indraraj26 replied:
I dont think so, setValidators will override the existing validator and clear the rest validator where as addValidators will add the validtor in existing one.

splincode replied:
why not `appendVaildators` or `putValidators` ?

sg-gdarnell replied:
`FormGroup` has `setControl` and `addControl` methods, so `setValidators`/`addValidators` would at least be consistent with that.

The documentation for those methods was recently updated to describe the difference between 'set' and 'add' in https://github.com/angular/angular/pull/42638

splincode replied:
oh, thank you)

dev054 replied:
It would be nice if you support `readonly ValidatorFn[]` in this signature (in fact, in all methods that deals with array parameters). Currently, this would fail:

```ts
const validators = [
    myValidator, myValidator2
] as const;
addValidators(validators);
              ~~~~~~~~~~
```

> Argument of type 'readonly [(control: AbstractControl) => ValidationErrors | null]' is not assignable to parameter of type 'ValidatorFn | ValidatorFn[]'.
  The type 'readonly [(control: AbstractControl) => ValidationErrors | null]' is 'readonly' and cannot be assigned to the mutable type 'ValidatorFn[]'.(2345

dylhunn replied:
@dev054 Since that affects all methods in the package, rather than just the proposed new ones, it would need to be changed consistently everywhere. Please feel free to open an issue about this!

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -251,7 +251,8 @@ export abstract class AbstractControl {
   set validator(validatorFn: ValidatorFn|null) {
-    this._rawValidators = this._composedValidatorFn = validatorFn;
+    this._composedValidatorFn = validatorFn;
+    this._rawValidators = validatorFn ? [validatorFn] : null;
```
It looks like we can use `makeValidatorsArray` helper method here.

AndrewKushnir replied:
Btw, do we need this conversion at all? If I understand correctly, newly added methods support the case when `_rawValidators` is a function.

dylhunn replied:
You're right, we don't actually need it.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -423,6 +425,11 @@ export abstract class AbstractControl {
     return this._updateOn ? this._updateOn : (this.parent ? this.parent.updateOn : 'change');
   }
 
+  private makeValidatorsArray<T extends ValidatorFn|AsyncValidatorFn>(v: T|T[]|null): T[]|null {
```
We should move this function outside of the class: these are no `this` references + a standalone function would be minified much better (vs the `makeValidatorsArray` name retained non-minified as a class method).

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -431,9 +438,9 @@ export abstract class AbstractControl {
-    this._rawValidators = newValidator;
-    this._composedValidatorFn = coerceToValidator(newValidator);
+  setValidators(validators: ValidatorFn|ValidatorFn[]|null): void {
+    this._rawValidators = this.makeValidatorsArray(validators);
+    this._composedValidatorFn = coerceToValidator(validators);
```
The `this.makeValidatorsArray` and the `coerceToValidator` functions kind of reverse each other (unnecessarily) in case `validators` is a `ValidatorFn` (a function). First we would turn that into an array and then convert it back to a function. Let's see if we can optimize that to avoid extra computation and memory allocation.

dylhunn replied:
I think you are misreading this? `coerceToValidator` is called on `validators`, not `this._rawValidators`, so in the event that `validators` is a singleton, it will do nothing.

dylhunn replied:
Resolved as discussed.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -444,9 +451,134 @@ export abstract class AbstractControl {
+   * @param validators The new validator function or functions to add to this control.
+   */
+  addValidators(validators: ValidatorFn|ValidatorFn[]): void {
+    if (!Array.isArray(validators)) {
+      validators = [validators];
+    }
```
We could probably reuse the `makeValidatorsArray` helper method inside this function as well (and simplify it a bit):

```
addValidators(validators: ValidatorFn|ValidatorFn[]): void {
  const validatorsArray = makeValidatorsArray(validators);
  const newValidators = validatorsArray.filter(v => !this.hasValidator(v));
  const currentValidators = makeValidatorsArray(this._rawValidators) ?? [];
  this.setValidators(currentValidators.push(...newValidators));
}
```

AndrewKushnir replied:
We could also extract common code even more (the naming inside the `mergeValidators` function can be improved :)):

```
// outside of the AbstractControl class:
function mergeValidators<T>(current: T|T[]|null, added: T|T[]|null): T[] {
  // or may be `makeValidatorsArray` should return an empty array instead of `null`?
  const _added = makeValidatorsArray(added) ?? [];
  const _current = makeValidatorsArray(current) ?? [];
  const _unique = _added.filter(v => !_current.includes(v));
  return _current.push(..._unique);
}

// inside the AbstractControl class:
addValidators(validators: ValidatorFn|ValidatorFn[]): void {
  const mergedValidators = mergeValidators<ValidatorFn>(this._rawValidators. validators);
  this.setValidators(mergedValidators);
}
```

You could use the same `mergeValidators` function for async validators too.

dylhunn replied:
I thought the idea was to avoid calling `setValidators` from inside `addValidators`? Also, won't this code unnecessarily copy the existing validators around?

dylhunn replied:
Resolved as discussed.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -444,9 +451,134 @@ export abstract class AbstractControl {
+   * @param validator The validator to check for presence. Compared by function reference.
+   * @returns Whether the provided validator was found on this control.
+   */
+  hasValidator(validator: ValidatorFn): boolean {
+    if (!Array.isArray(this._rawValidators)) {
+      return this._rawValidators === validator;
+    }
+    return !!this._rawValidators?.includes(validator);
+  }
```
Some common logic can also be extracted to a shared function and reused for sync and async validators:

```
  // outside of AbstractControl:
  function hasValidator<T>(validators: T|T[]|null, validator: T): boolean {
      return Array.isArray(validators)
        ? validators.includes(validator)
        : validators === validator; // TODO: see if that would be ok/expected for the `null` case
  }

  // inside of AbstractControl:
  hasValidator(validator: ValidatorFn): boolean {
    return hasValidator<ValidatorFn>(this._rawValidators, validator);
  }

  hasAsyncValidator(validator: AsyncValidatorFn): boolean {
    return hasValidator<AsyncValidatorFn>(this._rawAsyncValidators, validator);
  }
```

dylhunn replied:
Done, with the addition that the type param on hasValidator can be inferred.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -136,6 +136,14 @@ function isOptionsObj(validatorOrOpts?: ValidatorFn|ValidatorFn[]|AbstractContro
+  return Array.isArray(v) ? v : [v];
+}
+
+function hasValidator<T>(validators: T|T[]|null, validator: T): boolean {
```
For consistency, let's make sure arguments in both `makeValidatorsArray` and `hasValidator` function use the same format (either short of full name) and the generic types are defined in the same way.

AndrewKushnir replied:
Consider moving those helper methods to the `packages/forms/src/validators.ts` file.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -261,7 +269,8 @@ export abstract class AbstractControl {
   }
   set asyncValidator(asyncValidatorFn: AsyncValidatorFn|null) {
-    this._rawAsyncValidators = this._composedAsyncValidatorFn = asyncValidatorFn;
+    this._composedAsyncValidatorFn = asyncValidatorFn;
+    this._rawAsyncValidators = asyncValidatorFn ? [asyncValidatorFn] : null;
```
I think this change can be reverted (it was reverted for sync validators case).

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -444,9 +454,92 @@ export abstract class AbstractControl {
+   * @param validators The validator or validators to remove.
+   */
+  removeValidators(validators: ValidatorFn|ValidatorFn[]): void {
+    const newValidators = makeValidatorsArray(validators);
```
I think we should rename the const:
```suggestion
    const validatorsToRemove = makeValidatorsArray(validators);
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -404,8 +414,8 @@ export abstract class AbstractControl {
-   * A multicasting observable that emits an event every time the validation `status` of the control
-   * recalculates.
+   * A multicasting observable that emits an event every time the validation `status` of the
+   * control recalculates.
```
Could you plz revert formatting changes (here and in other places in this file) to keep the diff minimal?

dylhunn replied:
`prettierrc` runs as a precommit hook, so there is no easy way to stop the formatting.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -431,9 +429,9 @@ export abstract class AbstractControl {
-  setValidators(newValidator: ValidatorFn|ValidatorFn[]|null): void {
-    this._rawValidators = newValidator;
-    this._composedValidatorFn = coerceToValidator(newValidator);
+  setValidators(validators: ValidatorFn|ValidatorFn[]|null): void {
```
I think it'd be great to extend the docs for the `setValidators` and `setAsyncValidators` methods and refer to the `addValidators` and `addAsyncValidators` respectively:

```
  /**
   * Sets the synchronous validators that are active on this control.  Calling
   * this overwrites any existing sync validators.
   * 
   * If you want to add a new validator without affecting existing ones, consider
   * using `addValidators` method instead.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -444,9 +442,96 @@ export abstract class AbstractControl {
+  }
+
+  /**
+   * Add an asynchronous validator or validators to this control, without affecting other
```
nit: for consistency we should either use sync/async (short version) or synchronous/asynchronous (full version) in docs (I'd vote for full version).

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -444,9 +442,96 @@ export abstract class AbstractControl {
+   *
+   * Adding a validator that already exists will have no effect.
+   *
+   * @param validators The new async validator function or functions to add to this control.
```
Also "async" vs "asynchronous" here for consistency.

AndrewKushnir commented on packages/forms/src/validators.ts:
```diff
@@ -663,3 +663,27 @@ export function getControlAsyncValidators(control: AbstractControl): AsyncValida
+ * @param validator The validator to check.
+ * @returns Whether the validator is present.
+ */
+export function hasValidator<T>(validators: T|T[]|null, validator: T): boolean {
```
```suggestion
export function hasValidator<T extends ValidatorFn|AsyncValidatorFn>(validators: T|T[]|null, validator: T): boolean {
```

AndrewKushnir commented on packages/forms/test/form_control_spec.ts:
```diff
@@ -312,6 +312,56 @@ describe('FormControl', () => {
+    });
+
+    it('should check presence of and remove a validator set with addValidators', () => {
+      const c = new FormControl('');
```
nit: add a check that there is no validator present initially:
```suggestion
      const c = new FormControl('');
      expect(c.hasValidator(Validators.required)).toEqual(false);
      
```

AndrewKushnir commented on packages/forms/test/form_control_spec.ts:
```diff
@@ -312,6 +312,56 @@ describe('FormControl', () => {
+
+    it('should be able to remove a validator added multiple times', () => {
+      const c = new FormControl('', Validators.required);
+      c.addValidators(Validators.required);
+      c.addValidators(Validators.required);
+      expect(c.hasValidator(Validators.required)).toEqual(true);
```
I think we'd also need a separate test to verify that adding the same validator twice (or more) doesn't actually add multiple copies.

AndrewKushnir commented on packages/forms/test/form_control_spec.ts:
```diff
@@ -508,6 +558,61 @@ describe('FormControl', () => {
          expect(c.status).toEqual('DISABLED');
        }));
+
+    it('should check presence of and remove a validator set in the control constructor', () => {
```
We should also add extra tests for the `FormBuilder` class as well.

AndrewKushnir reviewed (approved):
@dylhunn thanks for addressing the feedback 👍
The changes look good, just added a few minor comments related to docs and tests.

AndrewKushnir commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -49,7 +49,7 @@
       "uncompressed": {
         "runtime-es2015": 1150,
-        "main-es2015": 162346,
+        "main-es2015": 163066,
```
Note for the `size-tracking` group reviewers: this increase is expected since there is a new non-tree-shakable code added in this PR.

AndrewKushnir replied:
I think this number should be slightly different now (since some common code was moved to helper functions). Could you please reset it back and rerun the CI to see if payload size is over the threshold?

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -430,27 +428,119 @@ export abstract class AbstractControl {
-  setValidators(newValidator: ValidatorFn|ValidatorFn[]|null): void {
-    this._rawValidators = newValidator;
-    this._composedValidatorFn = coerceToValidator(newValidator);
+  addValidators(validators: ValidatorFn|ValidatorFn[]): void {
+    const newValidators = makeValidatorsArray(validators).filter(v => !this.hasValidator(v));
+    const oldValidators = makeValidatorsArray(this._rawValidators);
+    this.setValidators(oldValidators.concat(newValidators));
   }
```
I think we can compress it a bit more to avoid duplication between sync and async validators logic:
```suggestion

  // outside of the AbstractControl class...
  function addValidators<T>(newValidators: T|T[], currentValidators: T|T[]|null): T[] {
    const _new = makeValidatorsArray(newValidators).filter(v => !hasValidator(currentValidators, v));
    const _current = makeValidatorsArray(currentValidators);
    return _current.concat(_new);
  }

  // inside the AbstractControl class...
  addValidators(validators: ValidatorFn|ValidatorFn[]): void {
    const newValidators = addValidators<ValidatorFn>(validators, this._rawValidators);
    this.setValidators(newValidators);
  }
  
  addAsyncValidators(validators: AsyncValidatorFn|AsyncValidatorFn[]): void {
    const newValidators = addValidators<AsyncValidatorFn>(validators, this._rawAsyncValidators);
    this.setValidators(newValidators);
  }
```

and similarly we can update remove validators function to extract common duplicated logic.

AndrewKushnir reviewed (commented):
@dylhunn FYI, just one more proposal to reduce the payload a bit.

alxhub reviewed (approved):
Reviewed-For: public-api

atscott reviewed (approved):
reviewed-for: public-api

IgorMinar commented on packages/forms/src/model.ts:
```diff
@@ -430,27 +428,115 @@ export abstract class AbstractControl {
    *
+   * If you want to add a new validator without affecting existing ones, consider
+   * using `addValidators` method instead.
+   */
```
can you please also update the `validator` and `asyncValidator` docs to clarify that those getters/setters expose the internal "composed" validatorFn, which all of the individually added/set validatorFns are part of.

IgorMinar reviewed (approved):
LGTM for the API changes! Thanks for all the docs 👏🏻 

Reviewed-for: public-api

IgorMinar reviewed (approved):
LGTM  for size changes, but you might want to consider removing the `setValidators`/`setAsyncValidators` methods since they are not strictly required (you can clear validators and readd them) and add a bit of JS. I recommend keeping the API surface as small as possible to reduce the long term maintenance/docs/payload size etc costs. The two methods don't make a huge difference in this case so I'll leave it up to you to consider the trade offs.

Reviewed-for: size-tracking

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: size-tracking

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -430,27 +428,115 @@ export abstract class AbstractControl {
+   *
+   * @param validators The validator or validators to remove.
+   */
+  removeValidators(validators: ValidatorFn|ValidatorFn[]): void {
+    const validatorsToRemove = makeValidatorsArray(validators);
+    const oldValidators = makeValidatorsArray(this._rawValidators);
+    this.setValidators(oldValidators.filter(v => !validatorsToRemove.includes(v)));
+  }
```
The `removeValidators` and `removeAsyncValidators` methods can also benefit from moving common logic outside of the AbstractControl class, similar to "add" methods:
```suggestion
  removeValidators(validators: ValidatorFn|ValidatorFn[]): void {
    this.setValidators(removeValidators(validators, this._rawValidators));
  }
```

petebacondarwin commented on packages/forms/src/model.ts:
```diff
@@ -430,27 +428,115 @@ export abstract class AbstractControl {
 
   /**
-   * Empties out the sync validator list.
+   * Empties out the synchronous validator list.
```
There are one or two other occurrences of "sync" in the docs in this file if you fancy cleaning them up.

dylhunn replied:
I went ahead and changed them!

petebacondarwin reviewed (approved):
Nice work @dylhunn.

Reviewed-for: public-api

---

There is a small typo in the commit message: `vaioud`. Also, personally I would limit the title of the commit to a shorter sentence, e.g. 

```
feat(forms): add hasValidators and addValidators methods (for both sync and async)
```

---

Finally there is a corner case that I think is a bit broken. Consider:

```
c.addValidators([Validators.required, Validators.required]);
```

In this case the validator is actually added twice and will be triggered twice.
It turns out that running `c.removeValidators(Validators.required)` will actually remove both instances, and `c.hasValidator(Validators.required)` also works as expected.
So perhaps it is not worth worrying about?

dylhunn commented:
> LGTM for size changes, but you might want to consider removing the `setValidators`/`setAsyncValidators` methods since they are not strictly required (you can clear validators and readd them) and add a bit of JS. I recommend keeping the API surface as small as possible to reduce the long term maintenance/docs/payload size etc costs. The two methods don't make a huge difference in this case so I'll leave it up to you to consider the trade offs.
> 
> Reviewed-for: size-tracking

Thanks for the comment! `setValidators`/`setAsyncValidators` are the two preexisting methods, so to remove them we'd have to deprecate and provide a migration schematic in v13, which seems out of scope for this PR. Let me know if I'm misunderstanding.

dylhunn commented:
> There is a small typo in the commit message: `vaioud`. Also, personally I would limit the title of the commit to a shorter sentence, e.g.

Fixed!

> Finally there is a corner case that I think is a bit broken. Consider:
> 
> ```
> c.addValidators([Validators.required, Validators.required]);
> ```

I went ahead and fixed this -- now, `c.addValidators([Validators.required, Validators.required])` will only add one copy.

dev054 commented:
@dylhunn it'd be nice to mention the addtion of `removeValidators` in the commit message too.

AndrewKushnir commented on packages/forms/src/validators.ts:
```diff
@@ -663,3 +663,41 @@ export function getControlAsyncValidators(control: AbstractControl): AsyncValida
+  return Array.isArray(validators) ? validators.includes(validator) : validators === validator;
+}
+
+export function addValidators<T extends ValidatorFn|AsyncValidatorFn>(
+    validators: T|T[], currentValidators: T|T[]|null): T[] {
+  const _new = [...new Set(makeValidatorsArray(validators))].filter(
+      v => !hasValidator(currentValidators, v));
+  const _current = makeValidatorsArray(currentValidators);
+  return _current.concat(_new);
+}
```
I think we can avoid creating extra intermediate sets/arrays, for example:
```suggestion
export function addValidators<T extends ValidatorFn|AsyncValidatorFn>(
    validators: T|T[], currentValidators: T|T[]|null): T[] {
  const current = makeValidatorsArray(currentValidators);
  const validatorsToAdd = makeValidatorsArray(validators);
  validatorsToAdd.forEach((v: T) => {
     // Note: if there are duplicate entries in the new validators array,
     // only the first one would be added to the current list of validarors.
     // Duplicate ones would be ignored since `hasValidator` would detect
     // the presence of a validator function and we update the current list in place.
     if (!hasValidator(current, v)) {
      current.push(v);
    }   
  });
  return current;
}
```
This case would be a bit less efficient for duplicate validators in the list, but would be much better from perf and memory consumption perspective for a regular case (i.e. we don't need to create extra sets and arrays).

AndrewKushnir replied:
Btw, we should also add tests for the case that Pete described (with having multiple validators in the input array).

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -245,7 +243,8 @@ export abstract class AbstractControl {
   /**
-   * The function that is used to determine the validity of this control synchronously.
+   * Returns the function that is used to determine the validity of this control synchronously.
+   * If multiple validators have been added, this will be a single composed function.
```
```suggestion
   * If multiple validators have been added, this will be a single composed function.
   * See `Validators.compose()` for additional information.
```

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -255,7 +254,8 @@ export abstract class AbstractControl {
   /**
-   * The function that is used to determine the validity of this control asynchronously.
+   * Returns the function that is used to determine the validity of this control asynchronously.
+   * If multiple validators have been added, this will be a single composed function.
```
```suggestion
   * If multiple validators have been added, this will be a single composed function.
   * See `Validators.composeAsync()` for additional information.
```

AndrewKushnir commented on packages/forms/src/validators.ts:
```diff
@@ -663,3 +663,41 @@ export function getControlAsyncValidators(control: AbstractControl): AsyncValida
+  return Array.isArray(validators) ? validators.includes(validator) : validators === validator;
+}
+
+export function addValidators<T extends ValidatorFn|AsyncValidatorFn>(
```
I think there is a value of adding JSDoc to this function as well, pointing out the behavior with duplicate validators.

AndrewKushnir commented on packages/forms/src/model.ts:
```diff
@@ -425,32 +425,118 @@ export abstract class AbstractControl {
    * When you add or remove a validator at run time, you must call
    * `updateValueAndValidity()` for the new validation to take effect.
    *
+   * Adding a validator that already exists will have no effect.
+   *
```
May be add a quick note on the behavior that Pete pointed out (and similar note to the `addAsyncValidators` function):
```suggestion
   * Adding a validator that already exists will have no effect.
   *
   * If duplicate validator functions are present in the `validators` array, only the first instance
   * would be added to a form control.
```

IgorMinar commented:
> @IgorMinar Thanks for the comment! setValidators/setAsyncValidators are the two preexisting methods, so to remove them we'd have to deprecate and provide a migration schematic in v13, which seems out of scope for this PR. Let me know if I'm misunderstanding.

@dylhunn ah, I see. I misread the diff. Removing an already existing API comes with a non-negligible cost, so I recommend against it in this case. Let's keep it as is in the PR. thank you

khodabandelu commented:
hy guys ,
this feature support custom validator or only common validator that exist in angular?
@dylhunn

AndrewKushnir commented:
> this feature support custom validator or only common validator that exist in angular?

@khodabandelu extra APIs added in this PR support any validators (built-in as well as custom ones).

e-oz commented:
Delicious! Thank you very much for this 👍

e-oz commented:
Is it possible to backport it to 12.2.x? Please!

JoostK commented:
@e-oz This is available in 12.2 and will not be backported further as it's a feature.

e-oz commented:
@JoostK then it's just not mentioned in the changelog :) Thank you for the good news 👍

hiepxanh commented:
![image](https://user-images.githubusercontent.com/16970990/130754194-276c6366-2168-4eb0-9800-b856358b68ce.png)
can you add this properties on this abstractcontrol?
I have this problem which cannot solve:
`Type 'AbstractControl' is missing the following properties from type 'FormControl': registerOnChange, registerOnDisabledChange, _applyFormState`
